### DISCUSSION TITLE
Oqtane UI Expert Help on Contract

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#2055] [General]
Hi Oqtane Community,

I know this is not a jobs board, but.... 

We are starting our 4th Oqtane app and 6 Blazor app this week.  We have settled on what we are calling the "BOBStack" for lack of a better term for Blazor, Oqtane & Blazorise.  We love Oqtane and have standardized on it as our web-app platform moving forward.  Two pain-points that plague us are A- the complexity around integrating 3rd party UI packages, and  B- the designing and building of mobile-ready themes.  I have consistent (but not full-time) work for someone who is familiar with the nuance of Oqtane when it comes building UI themes and integrating UI Controls like Blazorise.  We are trying to "up our game" to build industry leading UI, but our UI focused devs get a bit lost with the additional steps needed to properly debug and deploy in Oqtane.  The web tools they are used to using are just a bit different.  It's a specific skill, and that's why I'm asking here.

If you are proficient with Oqtane Themes, and also have the UI chops to implement great design given stellar design assets, I'd love to talk to you.

Regards,
Mark Davis
CTO Trifoia

**Comment by sbwalker** [comment]
@markdav-is you were an early adopter of Blazor and Oqtane so its great to hear about your experiences.

In regards to 3rd party UI component libraries I have a philosophical question for you. If we take a step back and think about WebForms - it was a framework which provided a rich set of UI components. However one of the criticisms of WebForms was that it was "an abstraction over the web" which actually contributed to developers not learning the base web semantics of HTML, CSS, and JavaScript, as they became fully dependent on the UI component abstractions. In 2009 Microsoft introduced MVC - which did not have a robust component model - and developers were forced to become familiar with the base semantics of the web (ie. HTML, CSS, JavaScript). Now we have Blazor... and although it is fully possible to use the base web semantics, it seems that developers want to use higher level component library abstractions. I am not sure if this is because they do not understand Blazor and assume that they need to use these higher level component library abstractions, or if they are just not comfortable with HTML, CSS, JavaScript and need the abstraction. You mention that your "UI focused devs get a bit lost" - and I am wondering if this is because design professionals are used to working directly with HTML, CSS, JavaScript - so UI component libraries would be an annoying abstraction for them? There are obviously both pros and cons to using UI component libraries so I am trying to understand the dynamics at work here. Note that this is not just a question for Oqtane but for the broader Microsoft ecosystem in general.

**Reply by markdav-is** [reply]
One of the benefits of Oqtane and Blazor as opposed to other JS-SPAs is the simplicity and developer efficiency.  If we are building business-oriented apps that have a high level of commonality there's a benefit in adopting an existing component library.  I'll use a specific example of data grids and mobile use.  Boring business apps have lots of CRUD and data grids that need to morph into cards on mobile.  They all get requirements for sorting and filtering.  given the option to solve this in-house or adopt a 3rd party package, I'm opting to buy vs build with the idea that my devs coming from .Net MVC and not JS-SPA will have less of a ramp.  I'm also playing into the "don't build frameworks" concept avoid building out own set of internal components.  The module nature of Oqtane allows me to compose pages, use templates and themes, provide navigation.  This is top-down support of concepts in [atomic-design](https://bradfrost.com/blog/post/atomic-web-design/).  the component libraries give us some bottom-up support buy providing composite controls as well as that abstraction layer you mentioned over the raw HTML/CSS.  I've found that the amount of time and effort devs will put into making custom controls turns into a rabbit-hole pretty quickly unless they work closely with the designers and have a firm grip on how to make things morph to various screen sizes.  For my use case, I want to lean on as much as I can from Oqtane and component-vendors to build these apps quickly, we as much flexability as possible.  With Oqtane I can engage more the production team (composing pages) and have the devs focus more on business rules and less on component construction.  Because we are ultimately getting paid to build a mirror of the business need and not a set of reusable components.  Where "UI focused devs get a bit lost" is adding a nuget package to an Oqtane app is not like adding a Nuget package to a simple blazor app.  We often have to un-zip the package and extract the staticwebcontent folder for the components to work properly.  This extra level of understanding is picked-up over time, but it take a while to take a Blazor dev and upskill them to being an Oqtane dev.  Love this conversation and would enjoy having it in person sometime.

**Reply by sbwalker** [reply]
@markdav-is thank you for your detailed answer... you provided some great insight into the business rationale of why many developers are choosing to use UI component libraries in Blazor. 

In regards to the complexity of developing for Oqtane it is important to keep in mind that Oqtane Modules are not stand-alone Blazor applications - they are actually Razor Class Libraries (RCLs)... so essentially they are no different than any UI component library (as these are RCLs as well). RCLs are developed independently and need to be deployed into a Blazor application. For officially packaged versions this is usually done via Nuget however when developing an RCL this is usually done by copying the assets manually for expediency in the development workflow. The complexity is that you need to make sure that all dependencies of the RCL are deployed as well. Note that this is not necessary if all of the dependencies already exist in the Blazor application already. For example if your RCL requires EF Core but EF Core is already being used in the Blazor application where the RCL is being deployed then you do not need to deploy EF Core as part of the RCL. It is for this reason that I suggested in another thread that it may be useful to have UI component library Modules which can install all the dependencies into an Oqtane instance - so that a Module developer does not need to manage these dependencies in their own Module deployment.

**Reply by markdav-is** [reply]
@sbwalker I'll check out that other thread about RCLs and library Modules.  I'll have to investigate how static assets get deployed from RCLs as that's new to me.  I really like the idea of building an Oqtane Library Module (OLM?) that wraps the 3rd party control.  That's a good bit of open-source volunteerism and would help alleviate that developer pain-point of extracting static assets.  Thanks for the conversation, it has me thinking about this in new ways.

**Reply by stsrki** [reply]
I've been following this thread and I think I can answer this question.

Static files in RCL are generated and are contained in a NuGet package. When those NuGets are included in the app and when the app is published it will also generate them under the `wwwroot` so that they can be accessed from the `./_content` subfolder. The main issue here is that most Blazor libraries require to manually define their own static JS and CSS files in the `index.html` or `_Host.cshtml`. So if any Oqtane module is created and imported it would also require to manually defining those files.

We at Blazorise have (partially) solved this problem by utilizing JS Modules. So we import JS files dynamically at runtime, instead of having them manually defined in `index.html` or `_Host.cshtml`. For CSS files we still don't have a good solution and so CSS files are still required to be defined manually.

**Reply by sbwalker** [reply]
@strski thank you for providing your feedback. Oqtane supports deploying static assets in the standard _content location. It also has run-time support (ie. without requiring any changes to _Host) for CSS files via its own Resources concept - and these resources can exist in the _content folder. So it's really not very complicated to support third party UI component libraries. The challenge is that a developer needs to understand where these assets need to be located in the runtime environment (which Visual Studio and Nuget obfuscates from the developer in a standard Blazor app).

**Reply by markdav-is** [reply]
@stsrki Thanks for chiming in and the splendid work on blazorise.  I didn't realize that that the JS modules were loading dynamically.  I'll see about removing them from pre-loaded resources and see if nothing breaks.  Because I'm following both projects it might make sense for me to build the Blazorise Oqtane Library Module and keep it up to date with the changes to each of these fast-moving projects.  That would be a boon for my own teams as well as the community at large.
